ldms/src/sampler/zfs_topvdevs/zfs_topvdevs.c

```diff
@@ -0,0 +1,474 @@
+#include <unistd.h>
+#include "ldms.h"
+#include "ldmsd.h"
+#include "config.h"
+#include "sampler_base.h"
+#include "zfs_common.h"
+#include <stddef.h>
+
+#define SAMP "zfs_topvdevs"
+
+/* Function prototypes */
+
+int get_stats(zpool_handle_t *, void *);
+static int vdevs_count(zpool_handle_t *, void *);
+typedef int (*stat_printer_f) (nvlist_t *, const char *, zpool_handle_t *);
+
+static ldmsd_msg_log_f log_fn;
+static base_data_t sampler_base;
+static libzfs_handle_t *g_zfs;
+
+static struct {
+	int vdev_list_idx;
+	int vdev_rec_idx;
+} index_store;
+
+static ldms_mval_t list_handle;
+static uint_t top_vdev_count;
+
+/* metric templates for a virtual device */
+static struct ldms_metric_template_s zfs_topvdevs[] = {
+	{"zpoolname", 0, LDMS_V_CHAR_ARRAY, "", MAX_ZFS_NAME_LEN},
+	{"state", 0, LDMS_V_CHAR_ARRAY, "", MAX_ZFS_NAME_LEN},
+	{"vdevname", 0, LDMS_V_CHAR_ARRAY, "", MAX_ZFS_NAME_LEN},
+	{"alloc", 0, LDMS_V_U64, "", 1},
+	{"free", 0, LDMS_V_U64, "", 1},
+	{"size", 0, LDMS_V_U64, "", 1},
+	{"read_bytes", 0, LDMS_V_U64, "", 1},
+	{"read_errors", 0, LDMS_V_U64, "", 1},
+	{"read_ops", 0, LDMS_V_U64, "", 1},
+	{"write_bytes", 0, LDMS_V_U64, "", 1},
+	{"write_errors", 0, LDMS_V_U64, "", 1},
+	{"write_ops", 0, LDMS_V_U64, "", 1},
+	{"checksum_errors", 0, LDMS_V_U64, "", 1},
+	{"fragmentation", 0, LDMS_V_U64, "", 1},
+	{"init_errors", 0, LDMS_V_U64, "", 1},
+	{"trim_errors", 0, LDMS_V_U64, "", 1},
+	{"slow_ios", 0, LDMS_V_U64, "", 1},
+	{0},
+};
+
+/* need to find a better way more intuitive than that
+ * to manage heap. Like auto resize as a base function */
+
+#define VDEV_METRICS_LEN (ARRAY_LEN(zfs_topvdevs) - 1)
+static int vdev_metric_ids[VDEV_METRICS_LEN];
+static size_t zpool_vdev_heap_sz;
+
+/*****************************************************************************
+ * Initialize the structure as schema and add them to the base schema.
+ * Also calculate the size of memory needed per schema and add it to the ldms
+ * schema list.
+ ****************************************************************************/
+
+static int initialize_ldms_structs()
+{
+	ldms_record_t zpool_vdev_def;	/* a pointer */
+	int rc = 0;
+
+	log_fn(LDMSD_LDEBUG, SAMP " initialize()\n");
+
+	/* Create the schema */
+	base_schema_new(sampler_base);
+	if (sampler_base->schema == NULL)
+		goto err1;
+
+	/* create the vdev record */
+	zpool_vdev_def = ldms_record_from_template("zfs_topvdevs_stats",
+						   zfs_topvdevs,
+						   vdev_metric_ids);
+	if (zpool_vdev_def == NULL)
+		goto err2;
+
+	zpool_vdev_heap_sz = ldms_record_heap_size_get(zpool_vdev_def);
+	rc = ldms_schema_record_add(sampler_base->schema, zpool_vdev_def);
+	if (rc < 0)
+		goto err3;
+
+	index_store.vdev_rec_idx = rc;
+	top_vdev_count = 0;
+	rc = zpool_iter(g_zfs, vdevs_count, NULL);
+	/* add error for iter in case here */
+	rc = ldms_schema_metric_list_add(sampler_base->schema,
+					 "zpool_vdev_list",
+					 NULL,
+					 top_vdev_count * zpool_vdev_heap_sz);
+	if (rc < 0)
+		goto err2;
```

Comment:
Should be `err3`?  And the next case below.

Reply:
same as for leaf. nope, if you read the comment, we delete the record manually only if it wasn't added to the metric. In that case it was done already.